Readme
EZSMT+ is an constraint answer set programming solver extended from EZSMT. It takes a program written in EZCSP language as input, translates it into formulas written in SMT-LIB, and calls SMT solver to compute answer sets.


I. Installation
To compile EZSMT+ use the command lines:
1. % ./configure
2. % make

You need to have the following tools in order to run EZSMT+:
1. gringo(https://sourceforge.net/projects/potassco/files/gringo/4.5.3/). Place the executable file called gringo into the EZSMT+/tools folder.
2. SMT solvers:
  - CVC4(http://cvc4.cs.stanford.edu/downloads/builds/x86_64-linux-opt/cvc4-1.4-x86_64-linux-opt) Place the executable file into the EZSMT+/tools folder, and rename it as cvc4. Enter the command line: 
        % chmod +x cvc4
  - Z3(https://github.com/Z3Prover/z3/releases/tag/z3-4.5.0) Download the pre-compiled version z3-4.5.0-xxx, unzip and place the executable file bin/z3 into the folder EZSMT+/tools
  - YICES(http://yices.csl.sri.com/releases/2.5.4/yices-2.5.4-x86_64-pc-linux-gnu-static-gmp.tar.gz) Download the binaries and place the executable file yices/bin/yices-smt2 into the folder EZSMT+/tools
3. A pre-compiled version of EZCSP pre-parser comes with the distribution in the tools directory. If it does not work, download and compile binary files from www.mbal.tk/ezcsp/index.html. Replace the pre-compiled pre-parser with the resulting one.



Please append the following content to a file in your home directory called .bashrc:
   export EZSMTPLUS="<full-path-to-ezsmt+-directory>"
where <full-path-to-ezsmt+-directory> stands for the full path to the EZSMT+ directory. 
   e.g. export EZSMTPLUS="/home/da/ezsmtPlus"

Then, running the command
  % source ~/.bashrc
will set up the needed environment variables.



We have tested the following versions:
- gringo 4.5.4 and 3.0.5
- EZCSP 2.0.0
- cvc4 1.4
- z3 4.5.1
- yices 2.5.4



II. Usage

EZSMT+ is invoked as follows:
   % ./ezsmtPlus -file <path> [-file <path>] [-file <path>] <num1> <num2> [-cvc4|-yices|-z3] [-PrintExtAS] [-levelRanking|-levelRankingStrong|-SCClevelRanking|-SCClevelRankingStrong] [-reducedCompletion] [-minimalUpperBound] [-non-linear]


where:
<path> is the path to the input program in ezcsp language
<num1> is an optional number of answer sets to be computed. 0 stands for all and 1 is default.
<num2> is an optional number of extended answer sets to be computed. 0 stands for all and 1 is default.
[-cvc4|-yices|-z3] instructs EZSMT+ to call the corresponding SMT solver. By default, -cvc4 is chosen.
[-PrintExtAS] instructs EZSMT+ to print out extended answer sets. By default, only answer sets will be printed.
[-levelRanking|-levelRankingStrong|-SCClevelRanking|-SCClevelRankingStrong] selects the typy of level ranking formulas produced for non-tight programs. By default, -SCClevelRanking is chosen.
[-reducedCompletion] will instruct EZSMT+ to remove the part of Clark's completion that is captured by a level ranking formula.
[-minimalUpperBound] sets the upper bound of level ranking variables to the number of atoms inside the Strongly Connected Component containing that variable. A bigger upper bound (the total number of atoms) would be selected by default.
[-non-linear] selects non-linear logics for SMT solvers.


III. Example
Consider a non-tight EZ program ex1.ez:
   cspdomain(fd).
   cspvar(x).
   a :- c, not b.
   b :- d, not a.
   c :- d, not b.
   c :- d, not a.
   d :- c, not a.
   d :- not e, not b.
   e :- a, b.
   e :- not a, not b.
   required(x>=10) :- e.
   required(x<=9) :- not e.
Running the command
   % ./ezsmtPlus -file ex1.ez 0 -yices
gives us the output:
   ......
   Number of answer sets: 2
   Answer: 1
    Answer set: e
   Answer: 2
    Answer set: c d a

IV. Benchmarks
We also provide a collection of benchmarks in the folder called benchmarks.

Three benchmarks, namely, Reverse Folding, Incremental Scheduling, and Weighted Sequence, come from the Third Answer Set Programming Competition (Calimeri et al. 2011). We include a benchmark called Blending problem (Biavaschi 2017) and extend it to Blending_mixed, which contains variables over both integers and reals. Also, we use Bouncing Ball domain (Bartholomew 2016). It is important to remark that the encoding for BB domain results in a tight program. This domain uses nonlinear constraints over real numbers. In such a case ezsmt + is unable to handle nontight programs. Yet, it can process (non-)tight programs with integer nonlinear constraints (utilizing z3 or yices) or (non-)tight programs with linear constraints over real or mixed domain (using z3 or cvc4). Three more benchmarks, namely, RoutingMin, RoutingMax, and Travelling Salesperson are obtained from Liu, Janhunen, and Niemela (2012). The original TS benchmark is an optimization problem, and we turn it into a decision one. The Labyrinth benchmark is extended from the domain presented in the Fifth Answer Set Programming Competition (Calimeri et al. 2016). This extension allows us to add integer linear constraints into the problem encoding. Next benchmark, Robotics, comes from Young, Balduccini, and Israney (2017). Also, we present results on two benchmarks from Balduccini et al. (2017), namely, Car and Generator.

References:
[2011] Calimeri, F.; Ianni, G.; Ricca, F.; and et al. 2011.
The third answer set programming competition: Pre-
liminary report of the system competition track. In
Proceedings of the International Conference on Logic
Programming and Nonmonotonic Reasoning (LPNMR),
388–403. Berlin, Heidelberg: Springer-Verlag.
[2017] Biavaschi, S. 2017. Automated reasoning methods
in hybrid systems. Annual Report of “Scuola Superiore
dell’Università di Udine”.
[2016] Bartholomew, M. 2016. Answer Set Programming
Modulo Theories. Ph.D. Dissertation, Arizona State
University.
[2012] Liu, G.; Janhunen, T.; and Niemela, I. 2012. An-
swer set programming via mixed integer programming.
In Knowledge Representation and Reasoning Conference.
2016] Calimeri, F.; Gebser, M.; Maratea, M.; and Ricca,
F. 2016. Design and results of the fifth answer set
programming competition. Artif. Intell. 231(C):151–
181.
[2017] Young, R.; Balduccini, M.; and Israney, A. 2017.
Casp for robot control in hybrid domains. In In Work-
shop on Answer Set Programming and Other Computing
Paradigms (ASPOCP17).
[2017] Balduccini, M.; Magazzeni, D.; Maratea, M.; and
Leblanc, E. C. 2017. Casp solutions for planning in
hybrid domains. Theory and Practice of Logic Program-
ming 17(4):591–633.

V. Questions?
Send an e-mail to Da Shen(dashen@unomaha.edu), Yuliya Lierler(ylierler@unomaha.edu)


changes
//3.57 Includes
//        HCF check        
//        modularity in model checking
//        loopformula by default
//3.58 fixes the bug with loopformula from now on for disj programs only part of loop formula is added where head atom is sat
//    
//3.59 includes -eloop flag that performs elementary set computation
// and adds loop formulas of elementary loops when needed
// 
//July 31st 2006: 
//3.60 tightness bug fix through away SCC of 1 element
//     now all the unsat loops are added 
//
//3.6x    Ilkka's circuit component's 
//        where implemented -bc flag is used plus bcircuit program
//        must be present  
//        impl is preliminary hence output is not readable
//        no positive time results were gained
//        as Cmodels CNF representation is more compact and seem to 
//        be better suit

// May 15, 2006
// -adding HCF verification functions for the whole program and for Sccs
// -fixing modular verification from DLV minTestUnf
// May 16, 2006
// - adding minTestNEWGnt, minTestNEWGntSCC that incorporate
//          some processing of a program before giving it to SAT
//          for verification  
// - adding  posBodyInScc, negBodyInM for shortening 
//          minTestNEWGntSCC minTestUnfSCC
// - fixing bug with minTestUnfSCC in loopformula building area
//   as we were trying to 1) build a loopformula  based on the assignment
//   instead of newly found model
//   and 2) old SCC could interfeer with new ones
//
//May 19,2006
// - fixinf bug with schur that occured with WF preprocessing
//   cleaning up code for WF so that it only contains functions for WF
//   itself
//
// - removed nonmodular dlv and newGnt
// - debuging of modular dlv method as it finds some component Unf although it is probably found
// 
//May 23, 2006
//
// -at ctable.cc editing (1*) so that if there are many solutions to
// find and program is tight and mchaff is used we say that program 
// is not tight and force system to work in assat way
// since there occured a very strange bug, i.e. mchaff would find correct
solutions, while cmodels would fail to print them out correctly.
three days ago bug did not occur, meanwhile the changes that happend look not
to be relevant to here. Open question!

May 24, 2006
- Mchaff has some strange error at the execution of stratcomp20.3 
it only finds one solution. I suspect it can have sth to do with the bug from
yesterday as mchaff itself finds the right solution, but cmodels has
troubles interpreting it right. As relsat uses the same algo everywhere.
I discontinue for now support of Chaff by not allowing it in Main to read as an argument. Since we have Zchaff it must be a good substitution.
// goals: mChaff executable is no longer publically available 
//               for now it is a question if mchaff should be supported
//               or we remove support of mchaff
//               zChaff is default  

May 25, 2006

- api and 
- cmodels 
now can be used as libraries, where api can read in the program
while cmodels may perform the computation

Tomorrow I plan to add the ruletype nested and store all the rules
as nested ones, this way i can reduce all the code twice as i will need
only one copy since all the operations for disjunctive and nondisjunctive
rules look to be the same.

important
-- adding new atom for each body this was we will have compact loopformulas
   with the atoms that occured
-- hcf
-- elementary loops
-- niemela's addition Analyse first before proceeding!


May 26:

  Added NestedRule class and now at the point of adopting the whole
code not to Basicrule basic but to Nestedrule
 
  At the moment i am at adopting createCompletion()

  Also extended class Clauses so that building clauses would be much handier


May 30

-  how about if all the heads of disjunctive rule are false-atoms:
 now we do not take care of such case

main.cc
/*
 * File main.cc 
 * Last modified on 2 19:34 2002
 * By Yuliya Babovich 
 *
 */

// Copyright 1998 by Patrik Simons
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston,
// MA 02111-1307, USA.
//
// Patrik.Simons@hut.fi

//3.53 version Aug 24, 05 removed hend1,pend1,nend1 from the rules
//that used to be copies of hen,pend,nend
//
//3.54 version Aug 25, 05:
//(1) bug at buildingLoopFormulas fixed so it could cause 
//problems for nontight programs
//(2) bug with rules ...v A v...:-..A... not being remove is fixed
//3.55 version Sep 21, 05:
//(1) Zchaff version 2004.11.15 is incorporated in place of Zchaff 2003
//(2) Cmodels is made compatable with gcc3.4/4.* thanks too Wolfgang Faber
//3.57 Includes
//        HCF check        
//        modularity in model checking
//        loopformula by default
//3.58 fixes the bug with loopformula from now on for disj programs only part of loop formula is added where head atom is sat
//    
//3.59 includes -eloop flag that performs elementary set computation
// and adds loop formulas of elementary loops when needed
// 
//3.60 tightness bug fix through away SCC of 1 element
//     now all the unsat loops are added 
//
//3.6x    Ilkka's circuit component's 
//        where implemented -bc flag is used plus bcircuit program
//        must be present  
//        impl is preliminary hence output is not readable
//        no positive time results were gained
//        as Cmodels CNF representation is more compact and seem to 
//        be better suit
//3.63 Datastructures added to support linear time WFM, and Reduct
//
//3.64-65 Bug fixes at WFM
//
//3.66 Bug fixes Zchaff to be portable into 64 bit archetecture
// Thanks to Yinlei Yu: 
// zchaff_dbase.cpp, line 324. Replacing "int displacement" with
//"long long displacement"
//3.67 Bug fixes Zchaff to be portable into 64 bit archetecture
// Thanks to Andre Neumann: on 64 bit archetecture UNSAT when SAT, due to no 
// return true; in the end of createClauses 
// zchaff_dbase.cpp, line 324. Replacing "int displacement" with
//"long long displacement"
// 
// 3.68 Esra's bug fix that could occur on constraint rules translation
// where new_atom was created instead of a false atom 
// + improved simplification
// + Added flag -ms that allows calling MINISAT
//    Minisat is now incorporated into Cmodels,
//    its code has been added as part of Cmodels
//    two functions added to SimpSolver to allow
//    adding clauses from cmodels, and getting feedback from a solver
//    once solution is found
//    Minisat and Zchaff communication is alike 
// + bug fixes for choice rules as {k}:-k.
// + bug fix with Weights that were not allowed to drop below negative value 
// version 3.69 
// + replaced zChaff version from 2004 with zChaff 2007.3.12
// + added check on what addClause from Minisat returns, since when
//   it returns false the instance is UNSAT.
// + -zca flag added only for comparison reasons of incremental SAT solving and 
//   assat method
// cmodels 3.70
// Esra's bug Fix
// WFM bug
//void ConstraintRule::HeadInOneRule(Atom* at)
//void     WeightRule::HeadInOneRule(Atom* at){
//some of the atoms could have been assigned values by now
//for instance a:-1{falseassigned1, notassigned}
//in such case only notassigned should be assigned computetrue
//Such bug could not appear in Basic, choice or disj rules
//
//
//3.71 improvement of clausification by not inroducing aux atom
//when -a occurs in the body i.e, avoiding intoruding clauses for aux_1==-a 
//+fixing intro two aux atoms for the same room
//
//3.73 atmost is computed (wfs--complete); sorting is introduced to disregard
//  same bodies in rules, and also the rules that are subsummed by others
//
//3.74 bug fixes due to Benjamin Kaufmann
// (i) weightrule Atmost computation
// (ii) completionchoice 
// (iii) weight rule elimination inefficiency in translation is removed 
//      (compare worked unproperly) cmodels worked correctly but inefficiently
//3.75 bug fix due to  Christian Drescher drescher@cs.uni-potsdam.de
//(occured in 3.73 not previously and only for the case of 
//disjunctive programs i.e.
// a v b:-c would be represented as two clauses: -c v a; -c v b 
// in place of -c v a v b
//
//+ following Benjamin suggestion changed minisatSolver->solve(true,false))
//  to minisatSolver->solve(true,true)) that disables farther than first time
//  preprocessing in minisat
//
//3.76 bug fixed in loop formula construction due to not cleanscc
//3.77 bug fixed in loop formula construction due to not cleanscc
//
//Minisat version 1 is added -ms1
//-loopformula1 flag (just one loopformula at a time is added
// instead of all
//-temp flag in case of Minisat v 1(-ms1) learns and forgets loop formulas
//
//
//3.78 bug fixed in read.cc 
//int Read::addWeightRule (FILE *f) 
// now reading is done with add_head() 
// in place of add_head_repetition 
// bug was reported by Roland Kaminski 26 Feb Fri 
//(it could occur only on programs with weight rules)
//3.79 
// 1) bug fixed in cmodels.cc 
// bool Cmodels::rec_weight_rule(Weight totalweight, int sizeC, Atom* headC,
//		     unsigned long atleast, int counter_body)
//
//weightrules translation had a bug by assigning the same atom to 
//two different expressions (see comments in the code that specify what was changed)
//
// bug was reported by  Raphael Finkel  10 Mar 2009
//(it could occur only on programs with weight rules)
// 2) in zchaff_dbase.cc displacement given a new type
//    to support both 32 and 64 bit archetectures in uniform manner.
//    The change is suggested by  Gurucharan Huchachar 11 Mar 2009
//    Line 329 zchaff_dbase.cc :
//    ptrdiff_t displacement = _lit_pool_start - old_start;
//    For 64 bit archetecture
//3.80
// minisat is now default instead of zchaff
//
//3.81
// capability for incremental ASP is added; please see ctable.h file 
// that contains all the details of the interface
//3.82
// under construction
// idea to implement an additional propagation of minisat 1 on cmodels side 
// so that non-lazy approach were possible with EZ-CSP or other system if it needs to
// 3.82g -> h bug fix due to Benjamin Kaufmann clearLoop added at cmodels.cc 4847
// 3.83 per Marcello's request change in partial assignment return to external solver. Now it is more flexible which atoms external solver wants to see. See ctable.h for Docs.
// 3.84 bug fix reported by Peter Schueller. Occurred in addDenial cmodels.cc
// 3.86 introduced -cdimacs flag to allow ezsmt 
//      to use cmodles capability compute completion
// 3.86.1 fixed declarations in Minisat so that gcc compiler version 5+ compiles the code (due to Marcello Balduccini)  
//      introduced -file option to pass filename as a parameter for reading (due to Ben Susman) 

#include <iostream>
#include <new>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "ctable.h"
#include "timer.h"
#include "interpret.h"
#include <time.h>
#include <sstream>
#include <ctype.h>
#include <fstream>

using namespace std;

Ctable ctable;
bool Atom::change = false;
bool Atom::conflict = false;
void a_new_handler ()
{
  cerr << "operator new failed: out of memory" << endl;
  exit(23);
}
static void timeOutHandler(int sig)
{
  switch (sig) {

#ifndef _WIN32	// marcy
  case SIGXCPU:
#ifndef SHORT_OUT
    ctable.cmodels.output.PrintStats();
#else
    ctable.cmodels.output.PrintStats();
#endif
    break;
#endif
  default:
    fprintf(stderr, "Unknown event\n");
    break;
  }

  exit (22);
}


int main (int argc, char *argv[])
{


  Timer mainTimer;
  

  set_new_handler (&a_new_handler);
  bool error = false;
 



  strcpy(ctable.cmodels.param.cmodelsname, &argv[0][0]);

  strcpy(ctable.cmodels.param.config, "CONFIG");
  
  for (int c = 1; c < argc && !error; c++){
    if (c+1<argc){
      c=c+ctable.setSingleExecutionArgument(&argv[c][0],&argv[c+1][0]);
    }
    else
      c=c+ctable.setSingleExecutionArgument(&argv[c][0],NULL);
  }


  //if the timeout was set then we will set the function for timeout
  if(ctable.cmodels.param.timeout!=0)    
     mainTimer.SetTimeout(ctable.cmodels.param.timeout, timeOutHandler);


  //This is not possble at the moment unless we include
  //clausification of loop formulas using many clauses
  //  if(ctable.cmodels.sys == 3 & ctable.cmodels.le)
    //    ctable.cmodels.sys =6; //here we force zchaff 
  //with incremental learning to be our choice of a sat solver
  
	
  //we find the path to cmodels and assume that config file is located there
  if(strcmp(ctable.cmodels.param.config,"\0")==0){
    char path_to_config[100];
    int k=0; 
    path_to_config[k]='\0';
    int length = strlen(ctable.cmodels.param.cmodelsname);
    int l = length-1;
    while(l!=-1 && ctable.cmodels.param.cmodelsname[l]!='/'){
      l--;
    }
    if(l>=0){
      for( k= 0;k<=l;k++)
	path_to_config[k]= ctable.cmodels.param.cmodelsname[k];
      path_to_config[k]='\0';
    }
    strcat(path_to_config, "CONFIG");
    strcpy(ctable.cmodels.param.config, path_to_config);
  }
 
  if (error)
    {
      ctable.usage ();
      return 1;
    }
  if(ctable.cmodels.output.asparagus==STANDARD)
    cerr << "ezsmt version 2.0.0 (ezsmtPlus) Reading..."<<endl;;


  if(ctable.cmodels.param.numOfFiles==0){
	  ctable.usage ();
  	  exit(1);
  }

  //preparsing
  stringstream ss;
  ss.clear();
  ss.str("");
  ss<<"$EZSMTPLUS/tools/pre-parser ";
  for( int a = 0; a< ctable.cmodels.param.numOfFiles ; a++){
	  if ( access( ctable.cmodels.param.files[a], F_OK ) == -1 ){
		  cerr <<" *** Error: file "<<ctable.cmodels.param.files[a]<<" does not exist. ***"<<endl;
		  exit(1);
	  }

	  ss<<ctable.cmodels.param.files[a]<<" ";
  }
  ss<<" > "<<ctable.cmodels.param.file<<".preparsed";

  cerr<<endl<<"Running pre-parsing command: "<<ss.str().c_str()<<endl;;
  system(ss.str().c_str());

  //Check errors from preparsing output
  ostringstream outputFile;
  ss.clear();
  ss.str("");
  ss<<ctable.cmodels.param.file<<".preparsed";
  ifstream in_file(ss.str().c_str());
  outputFile <<in_file.rdbuf();
  string outputFilelStr= outputFile.str();
  in_file.close();
  istringstream iss(outputFilelStr);
  string line;
  while(getline(iss,line)){
  //if a error is read, output error message
    	if(line.find("SYNTAX ERROR")!=string::npos){
    	     cout << " *** Error during preparsing. See output file " <<ctable.cmodels.param.file<<".preparsed ***"<< endl;
    	     exit(1);
    	}
  }


  //grounding
  ss.clear();
  ss.str("");
  ss<<"$EZSMTPLUS/tools/gringo "<<ctable.cmodels.param.file<<".preparsed"<<" > "<<ctable.cmodels.param.file<<".grounded";
  cerr<<"Running grounding command: "<<ss.str().c_str()<<endl;;
  system(ss.str().c_str());


  //Check errors from grounding output
   ostringstream outputFile2;
   ss.clear();
   ss.str("");
   ss<<ctable.cmodels.param.file<<".grounded";
   ifstream in_file2(ss.str().c_str());
   outputFile2 <<in_file2.rdbuf();
   string outputFilelStr2= outputFile2.str();
   in_file2.close();
   istringstream iss2(outputFilelStr);
   while(getline(iss2,line)){
   //if a error is read, output error message
     	if(line.find("error: ")!=string::npos  || line.find("ERROR: (gringo): ")!=string::npos){
   	     cout << " *** Error during grounding. See output file " <<ctable.cmodels.param.file<<".grounded ***"<< endl;
   	     exit(1);
       }
   }


  strcat(ctable.cmodels.param.file, ".grounded");

  ctable.cmodels.output.timerAll.start();

  if(strlen(ctable.cmodels.param.file) > 0)
  {
      freopen(ctable.cmodels.param.file,"r",stdin);
  }

  int bad = ctable.read(stdin);  
  if(ctable.cmodels.output.asparagus==STANDARD)
	cerr << "done"<<endl;       
  if (bad)
    {
      cerr << "Error in input" << endl;
      return 1;
    }
  //removes some setting that might be not fitting
  //for some specific SAT solver
  ctable.cmodels.param.finish();

   ctable.calculate();
   /*	 

   //	  for DEBUGGING externals replace above line with the following comented code
  ctable.setSolver(MINISAT1);
  int numLits=ctable.getNumberGroundedAtoms();
  int* answerset_lits = new int[numLits];
  ctable.Initialize(answerset_lits, numLits);
  if (numLits!=-2){
	cerr<<"preporcessing is sufficient here";
	exit(23);
  }
  answerset_lits[0]=2;
  answerset_lits[1]=3;
  //  answerset_lits[2]=4;
  //answerset_lits[3]=5;
  //answerset_lits[4]=6;
  numLits=2;
  ctable.markExternallyConstrainedAtoms (answerset_lits, numLits);

  numLits=ctable.getNumberGroundedAtoms();
  bool* assignments= new bool [ctable.cmodels.program.number_of_atoms];
  ctable.Solve(answerset_lits, numLits);

  int inner_count=0;
	//by default assignment and constraint_lits is false

	for(int j=0; j<ctable.cmodels.program.cmodelsAtomsFromThisId; j++){
	  assignments[j]= false;
	}

	for (int i=0; i<numLits; i++){
	  for(int indA=inner_count; indA<ctable.cmodels.program.cmodelsAtomsFromThisId; indA++){
		inner_count++;
		if(answerset_lits[i]==ctable.cmodels.program.atoms[indA]->get_lparse_id()){
		  
		  assignments[indA]= true;		  

		  break;
		}
	  }
	}  
	for (long i=0;i<ctable.cmodels.program.cmodelsAtomsFromThisId;i++){
	  if(assignments[i]){
		if(strcmp("#noname#",	ctable.cmodels.program.atoms[i]->atom_name()))
			printf("%s ",	ctable.cmodels.program.atoms[i]->atom_name ());
		  
	   
      }
    }  
   */
  ctable.cmodels.output.timerAll.stop();
  ctable.cmodels.output.print();

  return 0;
}

